Added getppid as debugger detection (#1687)
OWASP/owasp-mstg: Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md

[needle][script_jailbreak-detection-bypass] > run
```

#### Using getppid

Applications on iOS can detect if they have been started by a debugger by checking their parent PID. Normally, an application is started by the [launchd](http://newosxbook.com/articles/Ch07.pdf) process, which is the first process running in the _user mode_ and has PID=1. However, if a debugger starts an application, we can observe that `getppid` returns a PID different than 1. This detection technique can be implemented in the following way:

```swift
func AmIBeingDebugged() -> Bool {
    return getppid() != 1

> }```

Reviewer: Unfortunately I cannot fix this line with an automatic suggestion. Could you please edit it so that the closing ` ``` ` is on the next line? That's the weak point of suggestions, dealing with ` ``` ` ^^

Author: Done

[needle][script_jailbreak-detection-bypass] > run
```

> 
> 

Reviewer: Not required
```suggestion
```

> #### Using getppid

Reviewer: General question: does this work on release builds? It'd be nice if you could name some apps know to be implementing this.

Author: Sure, it works on Release builds from XCODE and also on apps from AppStore. You can easily verify it by using iOS-deploy and then 'ps -efj'. Have a look at Airbnb and Goodreads but they don't crash on it. They seem to send this data to metric services. I found the use of _getppid_ on 151 apps from 1300.

> Applications on iOS can detect if they have been started by a debugger by checking their parent PID. Normally, an application is started by [launchd](http://newosxbook.com/articles/Ch07.pdf) process, which is the first process running in the *user mode* and has PID=1. However, if a debugger starts an application, we can observe that `getppid()` returns PID different than 1. The whole detection can be implemented in the following way:

Reviewer: ```suggestion
Applications on iOS can detect if they have been started by a debugger by checking their parent PID. Normally, an application is started by the [launchd](http://newosxbook.com/articles/Ch07.pdf) process, which is the first process running in the _user mode_ and has PID=1. However, if a debugger starts an application, we can observe that `getppid` returns a PID different than 1. This detection technique can be implemented in the following way:
```

> ```objective-c
> - (bool)AmIBeingDebugged{
>     int ppid = getppid();
>     return ppid != 1;
> }
> ```

Reviewer: What about directly the Swift version?
```suggestion
```swift
func AmIBeingDebugged() -> Bool {
    return getppid() != 1
}```
```

Author: Right, it's probably more convenient for developers to have Swift code

> 
> The listings below present shortened outputs of `getppid` and `ps aux`  for two different cases:
> 
> 1. An application has been opened normally from the SpringBoard. 
> 
>    getppid() returns 1, so *launchd* is the parent of our process. 
> 
>    ```
>    owasp-iPhone:~ root# ps aux
>    USER   PID   COMMAND
>    root   1    /sbin/launchd
>    mobile 1480 /var/containers/Bundle/Application/.../OWASP.app/OWASP
>    ```
> 
> 2. An application has been opened by a debugger, e.g. `ios-deploy --debug --bundle OWASP.app`
> 
>    getppid() returns 1491, so *debugserver* is the parent of our process.
> 
>    ```
>    owasp-iPhone:~ root# ps aux
>    USER   PID   COMMAND
>    mobile 1491 /Developer/usr/bin/debugserver --lockdown --launch=frontboard
>    mobile 1492 /var/containers/Bundle/Application/.../OWASP.app/OWASP
>    ```
> 
> 

Reviewer: I'd remove this because the above info should be enough after adding the new details.
```suggestion
```
